File: Algo/Leetcode/221MaximalSquare.cpp
// dp. O(mn) space + O(mn) time
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int row = matrix.size();
        if(row == 0) return 0;
        int col = matrix[0].size();
        if(col == 0) return 0;

        vector<vector<int>> dp(row, vector<int>(col));
        int max_side = 0;
        for(int i=0; i<row; i++){
            for(int j=0; j<col; j++){
                if(matrix[i][j] == '1'){
                    if(i==0 || j==0) dp[i][j] = 1;
                    else dp[i][j] = min(min(dp[i-1][j-1], dp[i-1][j]), dp[i][j-1]) + 1;
                } else
                    dp[i][j] = 0;
                max_side = max(max_side, dp[i][j]);
            }
        }
        return max_side*max_side;
    }
};

File: Algo/Leetcode/221MaximalSquare.py
class Solution(object):
    def maximalSquare(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        max_len = 0
        n = len(matrix)
        m = len(matrix[0])

        for i in range(n):
            for j in range(m):
                if matrix[i][j] == "0":
                    matrix[i][j] = 0
                else:
                    matrix[i][j] = 1
                    max_len = 1

        for i in range(1, n):
            for j in range(1, m):
                if matrix[i][j] == 1 and matrix[i - 1][j] > 0 and matrix[i][j - 1] > 0 and matrix[i - 1][j - 1] > 0:
                    matrix[i][j] = min(matrix[i - 1][j - 1], matrix[i][j - 1], matrix[i - 1][j]) + 1        # min of these three
                    max_len = max(max_len, matrix[i][j])

        return max_len ** 2